docs/guides/spmd_api.md

```diff
@@ -0,0 +1,429 @@
+mesh
+```
+
+<!-- #region id="SJcTXjGYhe5o" -->
+# Model Definition
+
+Define an example layer `DotReluDot`, which creates two parameters `W1` and `W2` for dot products, and does a `jax.nn.relu` in between. To use it in `pjit` efficiently, you can use APIs in `pjit` and `flax.core.meta` to annotate the parameters and intermediate variables correctly:
+
+* Use [`nn.with_partitioning`](https://flax.readthedocs.io/en/latest/api_reference/_autosummary/flax.linen.with_partitioning.html#flax.linen.with_partitioning) to decorate the initializer function when creating params `W1` and `W2`.
+
+* Use `pjit.with_sharding_constraint` to annotate intermediate variables like `y` and `z` to force a particular sharding pattern under pjit when the ideal constraint is known.
+
+  * This is optional, but sometimes can help auto SPMD to partition efficiently. In the example given here, the call is not required, because `pjit` will figure out the same sharding layout for `y` and `z` regardless.
+
+Calling these APIs annotates each dimension of your parameter and intermediate variables with an axis name, namely `'data'`, `'model'` or `None`. This refers to how this dimension should be sharded - across one of the device mesh dimensions, or not sharded at all.
+
+For example, if we define `W1`, which has shape `(x.shape[-1], self.depth)`, with `(None, 'model')`, the first dimension (of length `x.shape[-1]`) will be replicated across all devices, while the second dimension (of length `self.depth`) will be sharded over the `model` axis of the device mesh.
+<!-- #endregion -->
+
+```python id="YChzss7uwwHx"
+class DotReluDot(nn.Module):
+  depth: int
+  @nn.compact
+  def __call__(self, x):
+    W1 = self.param(
+        'W1', 
+        nn.with_partitioning(nn.initializers.xavier_normal(), (None, 'model')),
+        (x.shape[-1], self.depth))
+
+    y = jax.nn.relu(jnp.dot(x, W1))
+    # force a local sharding annotation.
+    y = with_sharding_constraint(y, PartitionSpec('data', 'model'))
+
+    W2 = self.param(
+        'W2', 
+        nn.with_partitioning(nn.initializers.xavier_normal(), ('model', None)),
+        (self.depth, x.shape[-1]))
+
+    z = jnp.dot(y, W2)
+    # force a local sharding annotation.
+    z = with_sharding_constraint(z, PartitionSpec('data', None))
+
+    # Return a tuple to conform with the API `nn.scan` as shown in the cell below.
+    return z, None
+
+```
+
+<!-- #region id="1gvf6_a_OeU3" -->
+Now define the `MLP` model as multiple layers of `DotReluDot`.
+
+Note that you can use `nn.scan` and other [lifted transforms](https://flax.readthedocs.io/en/latest/advanced_topics/lift.html) on the module. This will introduce another dimension on the params, and this dimension should be annotated in `metadata_params`. Since the parameters inside `DotReluDot` are already sharded along `model` axis, we don't need to partition multiple layers across `model` dimension here, and therefore denote it as `None`.
```

Comment:
```suggestion
Note that you can use `nn.scan` and other [lifted transforms](https://flax.readthedocs.io/en/latest/advanced_topics/lift.html) on the module. This will introduce another dimension for the params (the dimension over which `scan` is applied), and this dimension should be speficied using the `metadata_params` argument. Since the parameters inside `DotReluDot` are already sharded along `model` axis, we don't need to partition multiple layers across `model` dimension here, and therefore denote it as `None`.
```